appserver/deployment/dol/src/main/java/com/sun/enterprise/deployment/util/ResourceValidator.java

```diff
@@ -0,0 +1,696 @@
+import org.glassfish.api.deployment.DeployCommandParameters;
+import org.glassfish.api.deployment.DeploymentContext;
+import org.glassfish.api.event.EventListener;
+import org.glassfish.api.event.Events;
+import org.glassfish.deployment.common.DeploymentException;
+import org.glassfish.internal.deployment.Deployment;
+import org.glassfish.logging.annotation.LogMessageInfo;
+import org.jvnet.hk2.annotations.Service;
+import org.glassfish.resourcebase.resources.api.ResourceConstants;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.naming.InitialContext;
+import javax.naming.NamingException;
+import java.net.MalformedURLException;
+import java.util.*;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Created by Krishna Deepak on 6/9/17.
+ */
+@Service
+public class ResourceValidator implements EventListener, ResourceValidatorVisitor {
+
+    public static final Logger deplLogger = com.sun.enterprise.deployment.util.DOLUtils.deplLogger;
+
+    @LogMessageInfo(
+            message = "JNDI lookup failed for the resource with jndi name: {0}",
+            level = "SEVERE",
+            cause = "JNDI lookup for the specified resource failed.",
+            action = "Create the necessary object before deploying the application.",
+            comment = "For the method validateJNDIRefs of com.sun.enterprise.deployment.util.ResourceValidator."
+    )
+    private static final String RESOURCE_REF_JNDI_LOOKUP_FAILED = "AS-DEPLOYMENT-00026";
+
+    String target;
+
+    DeploymentContext dc;
+
+    Application application;
+
+    @Inject
+    private Events events;
+
+    @Inject
+    Domain domain;
+
+    JNDINamespace myNamespace;
+    
+    public void postConstruct() {
+        events.register(this);
+    }
+
+    @Override
+    public void event(Event event) {
+        if (event.is(Deployment.AFTER_APPLICATION_CLASSLOADER_CREATION)) {
+            dc = (DeploymentContext) event.hook();
+            application = dc.getModuleMetaData(Application.class);
+            DeployCommandParameters commandParams = dc.getCommandParameters(DeployCommandParameters.class);
+            target = commandParams.target;
+            if (application == null)
+                return;
+            parseResources();
+            processResources();
+        }
+    }
+
+    /**
+     * Parse all the resources and store them in a namespace before starting the validation.
+     */
+    private void parseResources() {
+        myNamespace = new JNDINamespace();
+        parseResources(application);
+        for (BundleDescriptor bd : application.getBundleDescriptorsOfType(DOLUtils.warType())) {
+            parseResources((JndiNameEnvironment) bd);
+        }
+        for (BundleDescriptor bd : application.getBundleDescriptorsOfType(DOLUtils.carType())) {
+            parseResources((JndiNameEnvironment) bd);
+        }
+        for (BundleDescriptor bd : application.getBundleDescriptorsOfType(DOLUtils.ejbType())) {
```

Comment:
Instead of having three different loops based on archive type, would it be better to run it once and get the type from `BundleDescriptor` and call relevant method based on the type?